test/api/route/messages_test.dart

```diff
@@ -436,4 +453,195 @@ void main() {
+      FakeApiConnection connection, {
+      required List<int> messages,
+      required UpdateMessageFlagsOp op,
+      required MessageFlag flag,
+      required Map<String, String> expected,
+    }) async {
+      connection.prepare(json:
+        UpdateMessageFlagsResult(messages: messages).toJson());
+      final result = await updateMessageFlags(connection,
+        messages: messages, op: op, flag: flag);
+      check(connection.lastRequest).isNotNull().isA<http.Request>()
+        ..method.equals('POST')
+        ..url.path.equals('/api/v1/messages/flags')
+        ..bodyFields.deepEquals(expected);
+      return result;
+    }
+
+    test('smoke', () {
+      return FakeApiConnection.with_((connection) async {
+        await checkUpdateMessageFlags(connection,
+          messages: [1, 2, 3],
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'messages': jsonEncode([1, 2, 3]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
+    });
+  });
+
+  group('updateMessageFlagsForNarrow', () {
+    Future<UpdateMessageFlagsForNarrowResult> checkUpdateMessageFlagsForNarrow(
+      FakeApiConnection connection, {
+      required Anchor anchor,
+      required int numBefore,
+      required int numAfter,
+      required ApiNarrow narrow,
+      required UpdateMessageFlagsOp op,
+      required MessageFlag flag,
+      required Map<String, String> expected,
+    }) async {
+      final result = await updateMessageFlagsForNarrow(connection,
+        anchor: anchor, numBefore: numBefore, numAfter: numAfter,
+        narrow: narrow, op: op, flag: flag);
+      check(connection.lastRequest).isNotNull().isA<http.Request>()
+        ..method.equals('POST')
+        ..url.path.equals('/api/v1/messages/flags/narrow')
+        ..bodyFields.deepEquals(expected);
+      return result;
+    }
+    final fakeResult = UpdateMessageFlagsForNarrowResult(
+      processedCount: 11, updatedCount: 11, foundOldest: true,
+      foundNewest: true);
+
+    test('smoke', () {
+      return FakeApiConnection.with_((connection) async {
+        connection.prepare(json: fakeResult.toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+          anchor: AnchorCode.oldest,
+          numBefore: 0, numAfter: 20,
+          narrow: const AllMessagesNarrow().apiEncode(),
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'anchor': 'oldest',
+            'num_before': '0',
+            'num_after': '20',
+            'narrow': jsonEncode([]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
+    });
+
+    test('numeric anchor', () {
+      return FakeApiConnection.with_((connection) async {
+        connection.prepare(json: fakeResult.toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+          anchor: const NumericAnchor(42),
+          numBefore: 0, numAfter: 20,
```

Comment:
nit: if `numBefore` is 0, then the real server behavior will be for `foundOldest` to be false (because it won't look to see if there are any older); so the test data should reflect that

(I'm not sure if that's still true when `anchor` is `oldest`, as above. Could do a quick experiment to find out. If not, then `fakeResult` could become a function taking `{required bool foundOldest}`.)

Reply:
Confirmed by both a quick experiment and looking in zerver source that `anchor=oldest` does send `true` for `foundOldest`.